src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright 2018- The Pixie Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.h"
+
+#include <algorithm>
+#include <deque>
+#include <map>
+#include <set>
+#include <string>
+#include <tuple>
+#include <utility>
+
+#include "src/common/json/json.h"
+#include "src/stirling/source_connectors/socket_tracer/protocols/mqtt/types.h"
+
+namespace px {
+namespace stirling {
+namespace protocols {
+namespace mqtt {
+
+typedef uint8_t MatchKey;
+
+constexpr MatchKey UnmatchedResp = 0xff;
+
+std::map<MatchKey, MatchKey> MapRequestToResponse = {
+    // CONNECT to CONNACK
+    {0x10, 0x20},
+    // PUBLISH QOS 0 to Dummy response
+    {0x30, UnmatchedResp},
+    {0x31, UnmatchedResp},
+    {0x38, UnmatchedResp},
+    {0x39, UnmatchedResp},
+    // PUBLISH QOS 1 to PUBACK
+    {0x32, 0x40},
+    {0x33, 0x40},
+    {0x3a, 0x40},
+    {0x3b, 0x40},
+    // PUBLISH QOS 2 to PUBREC
+    {0x34, 0x50},
+    {0x35, 0x50},
+    {0x3c, 0x50},
+    {0x3d, 0x50},
+    // PUBREL to PUBCOMP
+    {0x60, 0x70},
+    // SUBSCRIBE to SUBACK
+    {0X80, 0X90},
+    // UNSUBSCRIBE to UNSUBACK
+    {0xa0, 0xb0},
+    // PINGREQ to PINGRESP
+    {0xc0, 0xd0},
+    // DISCONNECT to Dummy response
+    {0xe0, UnmatchedResp},
+    // AUTH to Dummy response
+    {0xf0, UnmatchedResp}};
+
+std::set<std::tuple<uint32_t, uint32_t>> UnansweredPublish;
```

Comment:
I was wondering if when the server initiates PUBLISH, the message type is going to be a `kRequest` or a `kResponse`. I am assuming that it will be a `kResponse`, since in `conn_tracker.cc` and `conn_tracker.h` it looks like whether the type is passed as `kRequest` or `kResponse` depends on the role of the sender,
conn_tracker.cc,
```cpp
DataStream* ConnTracker::req_data() {
  DCHECK_NE(role_, kRoleUnknown);
  switch (role_) {
    case kRoleClient:
      return &send_data_;
    case kRoleServer:
      return &recv_data_;
    default:
      return nullptr;
  }
}
```
conn_tracker.h,
```cpp    
req_data_ptr->template ProcessBytesToFrames<TKey, TFrameType, TStateType>(
    message_type_t::kRequest, state_ptr);

DataStream* resp_data_ptr = resp_data();
DCHECK_NE(resp_data_ptr, nullptr);
resp_data_ptr->template ProcessBytesToFrames<TKey, TFrameType, TStateType>(
    message_type_t::kResponse, state_ptr);
```
This would help me put increment the counter on the right map in the state.

Other than that, this idea seems like a great solution for the duplicate PUBLISH problem. Thanks for pointing it out!

Reply:
I believe it's going to be a `kResponse` as well. Since this is the first protocol that Pixie will support that is bidirectional (client and server can send requests), we will need to verify that -- which is why I wanted to leave a TODO (from [this thread](https://github.com/pixie-io/pixie/pull/1918#discussion_r1643096915)). Our common heuristics (clearing all unmatched response frames) assume that the server won't send requests (labeled as responses).

No problem!